pkg/credentialprovider/azure/azure_credentials.go

```diff
@@ -177,6 +170,16 @@ func (a *acrProvider) Provide() credentialprovider.DockerConfig {
 	defer cancel()
 
 	glog.V(4).Infof("listing registries")
+
+	var err error
+	a.servicePrincipalToken, err = auth.GetServicePrincipalToken(a.config, a.environment)
```

Comment:
This will call AAD API every time to get token every time 'Provide' is called.
Can it cache 'servicePrincipalToken', and 'registryClient', and only do a re-fetch if later 'registryClient.List' fails?

Reply:
+1 to do this only when not getting the expected result. 

Since the list is actually successful but without the expected result, you can add a check of the list, e.g. `len(result) == 0`?

Reply:
@karataliu @feiskyer The `Provide` func call happens when there is pull image requirement, and if pull image fails, it will try to rerun `Provide` by `lookup` func every 5 min, so not every kubelet will call this func only when there is pull image failure, related invocation code is here:
https://github.com/kubernetes/kubernetes/blob/795b7da8b05d2ad39567b0fe70fb9d6390cb4d0b/pkg/kubelet/kuberuntime/kuberuntime_image.go#L45